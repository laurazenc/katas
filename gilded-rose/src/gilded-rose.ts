export class Item {
	name: string;
	sellIn: number;
	quality: number;

	constructor(name: string, sellIn: number, quality: number) {
		this.name = name;
		this.sellIn = sellIn;
		this.quality = quality;
	}
}

export class GildedRose {
	items: Array<Item>;

	constructor(items = [] as Array<Item>) {
		this.items = items;
	}

	updateQuality() {
		for (let i = 0; i < this.items.length; i++) {
			this.doUpdateQuality(this.items[i]);
		}

		return this.items;
	}

	private doUpdateQuality(item: Item) {
		switch (item.name) {
			case "Aged Brie":
				if (item.quality < 50) {
					item.quality = item.quality + 1;
				}
				item.sellIn = item.sellIn - 1;
				if (item.sellIn < 0) {
					if (item.quality < 50) {
						item.quality = item.quality + 1;
					}
				}
				break;
			case "Backstage passes to a TAFKAL80ETC concert":
				if (item.quality < 50) {
					item.quality = item.quality + 1;
					if (item.sellIn < 11) {
						if (item.quality < 50) {
							item.quality = item.quality + 1;
						}
					}
					if (item.sellIn < 6) {
						if (item.quality < 50) {
							item.quality = item.quality + 1;
						}
					}
				}
				item.sellIn = item.sellIn - 1;
				if (item.sellIn < 0) {
					item.quality = item.quality - item.quality;
				}
				break;
			case "Sulfuras, Hand of Ragnaros":
				break;
			default:
				if (item.quality > 0) {
					item.quality = item.quality - 1;
				}
				item.sellIn = item.sellIn - 1;
				if (item.sellIn < 0) {
					if (item.quality > 0) {
						item.quality = item.quality - 1;
					}
				}
				break;
		}
	}
}
