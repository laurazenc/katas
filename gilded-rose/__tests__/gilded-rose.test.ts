import { verifyAllCombinations3 } from "approvals/lib/Providers/Jest/CombinationApprovals";

import { GildedRose, Item } from "../src/gilded-rose.ts";

function calculateQuality(name: string, quality: number, sellIn: number) {
	const gildedRose = new GildedRose([new Item(name, sellIn, quality)]);
	const result = gildedRose.updateQuality();

	return JSON.stringify(result[0]);
}

describe("Gilded Rose", () => {
	it("should calculate quality", () => {
		const names = [
			"foo",
			"Sulfuras, Hand of Ragnaros",
			"Aged Brie",
			"Backstage passes to a TAFKAL80ETC concert",
		];
		const qualities = [0, 1, 2, 50];
		const sellIns = [-1, 0, 1, 11, 6];
		verifyAllCombinations3(calculateQuality, names, qualities, sellIns);
	});
});
