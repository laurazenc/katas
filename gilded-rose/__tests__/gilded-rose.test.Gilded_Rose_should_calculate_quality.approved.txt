[foo,0,-1] => {"name":"foo","sellIn":-2,"quality":0}
[foo,0,0] => {"name":"foo","sellIn":-1,"quality":0}
[foo,0,1] => {"name":"foo","sellIn":0,"quality":0}
[foo,0,11] => {"name":"foo","sellIn":10,"quality":0}
[foo,0,6] => {"name":"foo","sellIn":5,"quality":0}
[foo,1,-1] => {"name":"foo","sellIn":-2,"quality":0}
[foo,1,0] => {"name":"foo","sellIn":-1,"quality":0}
[foo,1,1] => {"name":"foo","sellIn":0,"quality":0}
[foo,1,11] => {"name":"foo","sellIn":10,"quality":0}
[foo,1,6] => {"name":"foo","sellIn":5,"quality":0}
[foo,2,-1] => {"name":"foo","sellIn":-2,"quality":0}
[foo,2,0] => {"name":"foo","sellIn":-1,"quality":0}
[foo,2,1] => {"name":"foo","sellIn":0,"quality":1}
[foo,2,11] => {"name":"foo","sellIn":10,"quality":1}
[foo,2,6] => {"name":"foo","sellIn":5,"quality":1}
[foo,50,-1] => {"name":"foo","sellIn":-2,"quality":48}
[foo,50,0] => {"name":"foo","sellIn":-1,"quality":48}
[foo,50,1] => {"name":"foo","sellIn":0,"quality":49}
[foo,50,11] => {"name":"foo","sellIn":10,"quality":49}
[foo,50,6] => {"name":"foo","sellIn":5,"quality":49}
[Sulfuras, Hand of Ragnaros,0,-1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":-1,"quality":0}
[Sulfuras, Hand of Ragnaros,0,0] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":0,"quality":0}
[Sulfuras, Hand of Ragnaros,0,1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":1,"quality":0}
[Sulfuras, Hand of Ragnaros,0,11] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":11,"quality":0}
[Sulfuras, Hand of Ragnaros,0,6] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":6,"quality":0}
[Sulfuras, Hand of Ragnaros,1,-1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":-1,"quality":1}
[Sulfuras, Hand of Ragnaros,1,0] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":0,"quality":1}
[Sulfuras, Hand of Ragnaros,1,1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":1,"quality":1}
[Sulfuras, Hand of Ragnaros,1,11] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":11,"quality":1}
[Sulfuras, Hand of Ragnaros,1,6] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":6,"quality":1}
[Sulfuras, Hand of Ragnaros,2,-1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":-1,"quality":2}
[Sulfuras, Hand of Ragnaros,2,0] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":0,"quality":2}
[Sulfuras, Hand of Ragnaros,2,1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":1,"quality":2}
[Sulfuras, Hand of Ragnaros,2,11] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":11,"quality":2}
[Sulfuras, Hand of Ragnaros,2,6] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":6,"quality":2}
[Sulfuras, Hand of Ragnaros,50,-1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":-1,"quality":50}
[Sulfuras, Hand of Ragnaros,50,0] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":0,"quality":50}
[Sulfuras, Hand of Ragnaros,50,1] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":1,"quality":50}
[Sulfuras, Hand of Ragnaros,50,11] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":11,"quality":50}
[Sulfuras, Hand of Ragnaros,50,6] => {"name":"Sulfuras, Hand of Ragnaros","sellIn":6,"quality":50}
[Aged Brie,0,-1] => {"name":"Aged Brie","sellIn":-2,"quality":2}
[Aged Brie,0,0] => {"name":"Aged Brie","sellIn":-1,"quality":2}
[Aged Brie,0,1] => {"name":"Aged Brie","sellIn":0,"quality":1}
[Aged Brie,0,11] => {"name":"Aged Brie","sellIn":10,"quality":1}
[Aged Brie,0,6] => {"name":"Aged Brie","sellIn":5,"quality":1}
[Aged Brie,1,-1] => {"name":"Aged Brie","sellIn":-2,"quality":3}
[Aged Brie,1,0] => {"name":"Aged Brie","sellIn":-1,"quality":3}
[Aged Brie,1,1] => {"name":"Aged Brie","sellIn":0,"quality":2}
[Aged Brie,1,11] => {"name":"Aged Brie","sellIn":10,"quality":2}
[Aged Brie,1,6] => {"name":"Aged Brie","sellIn":5,"quality":2}
[Aged Brie,2,-1] => {"name":"Aged Brie","sellIn":-2,"quality":4}
[Aged Brie,2,0] => {"name":"Aged Brie","sellIn":-1,"quality":4}
[Aged Brie,2,1] => {"name":"Aged Brie","sellIn":0,"quality":3}
[Aged Brie,2,11] => {"name":"Aged Brie","sellIn":10,"quality":3}
[Aged Brie,2,6] => {"name":"Aged Brie","sellIn":5,"quality":3}
[Aged Brie,50,-1] => {"name":"Aged Brie","sellIn":-2,"quality":50}
[Aged Brie,50,0] => {"name":"Aged Brie","sellIn":-1,"quality":50}
[Aged Brie,50,1] => {"name":"Aged Brie","sellIn":0,"quality":50}
[Aged Brie,50,11] => {"name":"Aged Brie","sellIn":10,"quality":50}
[Aged Brie,50,6] => {"name":"Aged Brie","sellIn":5,"quality":50}
[Backstage passes to a TAFKAL80ETC concert,0,-1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-2,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,0,0] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-1,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,0,1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":0,"quality":3}
[Backstage passes to a TAFKAL80ETC concert,0,11] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":10,"quality":1}
[Backstage passes to a TAFKAL80ETC concert,0,6] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":5,"quality":2}
[Backstage passes to a TAFKAL80ETC concert,1,-1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-2,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,1,0] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-1,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,1,1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":0,"quality":4}
[Backstage passes to a TAFKAL80ETC concert,1,11] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":10,"quality":2}
[Backstage passes to a TAFKAL80ETC concert,1,6] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":5,"quality":3}
[Backstage passes to a TAFKAL80ETC concert,2,-1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-2,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,2,0] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-1,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,2,1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":0,"quality":5}
[Backstage passes to a TAFKAL80ETC concert,2,11] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":10,"quality":3}
[Backstage passes to a TAFKAL80ETC concert,2,6] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":5,"quality":4}
[Backstage passes to a TAFKAL80ETC concert,50,-1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-2,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,50,0] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":-1,"quality":0}
[Backstage passes to a TAFKAL80ETC concert,50,1] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":0,"quality":50}
[Backstage passes to a TAFKAL80ETC concert,50,11] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":10,"quality":50}
[Backstage passes to a TAFKAL80ETC concert,50,6] => {"name":"Backstage passes to a TAFKAL80ETC concert","sellIn":5,"quality":50}
